### PR TITLE
feat(builder): support modifyBundlerChain

### PR DESCRIPTION
**Pull request by 9aoy**
# PR Details

add modifyBundlerChain api to reuse Plugin between webpack provider and rspack provider.
- support bunderChain (webpack-chain runtime + BunderChain type declare) 
- add modifyBundlerChain api
   - in webpack：modifyBundlerChain => modifyWebpackChain => modifyWebpackConfig
   - in rspack：modifyBundlerChain => modifyRspackConfig
- reuse plugin-devtool & plugin-target
- move webpack-5-chain from builder-webpack-provider to builder-shared

Not added in documentation because the api is still unstable.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] I have updated changeset
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: packages/builder/builder-shared/src/test-stub/helper.ts [hunk 1]
```diff
@@ -13,4 +13,15 @@ export const mockBuilderPlugins: Plugins = {
...
   startUrl: genMockPlugin('builder-plugin-start-url'),
   fileSize: genMockPlugin('builder-plugin-file-size'),
+  target: genMockPlugin('builder-plugin-target'),
+  devtool: () =>
+    Promise.resolve({
+      name: 'builder-plugin-devtool',
+      setup: api => {
+        api.modifyBundlerChain((chain: any) => {
+          // test modifyBundlerChain work correctly
+          chain.devtool('cheap-module-source-map');
```
Comment (chenjiahan) [inline packages/builder/builder-shared/src/test-stub/helper.ts:23]: 这个是临时代码吗

File: scripts/prebundle/src/constant.ts [hunk 2]
```diff
@@ -181,7 +181,7 @@ export const TASKS: TaskConfig[] = [
   {
     packageDir: 'builder/builder-shared',
     packageName: '@modern-js/builder-shared',
-    dependencies: ['open'],
+    dependencies: ['open', 'webpack-5-chain'],
```
Comment (chenjiahan) [inline scripts/prebundle/src/constant.ts:184]: tapable 打进去了？我看 compiled/webpack-5-chain 的文件没变化

Comment (9aoy) [reply scripts/prebundle/src/constant.ts:184]: 我看 packageJson 和打包后的产物都没有 tapable。确定是需要的？

File: packages/builder/builder-shared/src/test-stub/helper.ts (hunk 1 above)
Comment (9aoy) [reply packages/builder/builder-shared/src/test-stub/helper.ts:23]: 不是，test-stub 是给测试用的

File: scripts/prebundle/src/constant.ts (hunk 2 above)
Comment (chenjiahan) [reply scripts/prebundle/src/constant.ts:184]: 看了下 webpack-chain 是要的，webpack-5-chain 把这个依赖干掉了，那没事了

File: packages/builder/builder-shared/src/test-stub/helper.ts (hunk 1 above)
Comment (chenjiahan) [reply packages/builder/builder-shared/src/test-stub/helper.ts:23]: 这段逻辑放在 stub 里怪怪的，看不出有啥关联

Comment (chenjiahan) [reply packages/builder/builder-shared/src/test-stub/helper.ts:23]: 如果是测试 modifyBundlerChain 的可用性，感觉应该放在单测里？

Comment (9aoy) [reply packages/builder/builder-shared/src/test-stub/helper.ts:23]: 👌  我换种方式吧

Comment (9aoy) [reply packages/builder/builder-shared/src/test-stub/helper.ts]: done

**Review by chenjiahan — APPROVED** [review]

**Pull request merged**